File: app/src/main/java/com/mudrichenko/evgeniy/flickrtestproject/ui/profile/ProfilePresenter.java
package com.mudrichenko.evgeniy.flickrtestproject.ui.profile;

import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentManager;
import android.support.v4.app.FragmentStatePagerAdapter;
import android.support.v4.content.ContextCompat;
import android.support.v4.graphics.ColorUtils;
import android.support.v4.view.PagerAdapter;

import com.arellomobile.mvp.InjectViewState;
import com.arellomobile.mvp.MvpPresenter;
import com.mudrichenko.evgeniy.flickrtestproject.App;
import com.mudrichenko.evgeniy.flickrtestproject.R;
import com.mudrichenko.evgeniy.flickrtestproject.ui.albums.AlbumsFragment;
import com.mudrichenko.evgeniy.flickrtestproject.ui.cameraRoll.CameraRollFragment;
import com.mudrichenko.evgeniy.flickrtestproject.ui.contactList.ContactListFragment;
import com.mudrichenko.evgeniy.flickrtestproject.ui.publicPhotos.PublicPhotosFragment;

@InjectViewState
public class ProfilePresenter extends MvpPresenter<ProfileView> {

    private final static int NUM_OF_PAGES = 4;

    private final int CAMERA_ROLL_FRAGMENT_ID = 0;
    private final int PUBLIC_PHOTOS_FRAGMENT_ID = 1;
    private final int ALBUMS_FRAGMENT_ID = 2;
    private final int CONTACT_LIST_FRAGMENT_ID = 3;

    private final int DEFAULT_FRAGMENT_ID = 0;

    private int mViewPagerCurrentPosition;

    private int backgroundColorCameraRoll;
    private int backgroundColorPublicPhotos;
    private int backgroundColorAlbums;
    private int backgroundColorContactList;

    private CameraRollFragment mCameraRollFragment;
    private PublicPhotosFragment mPublicPhotosFragment;
    private AlbumsFragment mAlbumsFragment;
    private ContactListFragment mContactListFragment;

    public ProfilePresenter() {
        mViewPagerCurrentPosition = DEFAULT_FRAGMENT_ID;
        backgroundColorCameraRoll = ContextCompat.getColor(App.Companion.getAppContext(), R.color.backgroundCameraRoll);
        backgroundColorPublicPhotos = ContextCompat.getColor(App.Companion.getAppContext(), R.color.backgroundPublicPhotos);
        backgroundColorAlbums = ContextCompat.getColor(App.Companion.getAppContext(), R.color.backgroundAlbums);
        backgroundColorContactList = ContextCompat.getColor(App.Companion.getAppContext(), R.color.backgroundContactList);
    }

    public void setViewPagerCurrentPosition(int position) {
        mViewPagerCurrentPosition = position;
    }

    public int getViewPagerCurrentPosition() {
        return mViewPagerCurrentPosition;
    }

    public PagerAdapter getPagerAdapter(FragmentManager fragmentManager) {
        return new MyFragmentPagerAdapter(fragmentManager);
    }

    public void onViewPagerScrolled(int position, float positionOffset) {
        if (position == 0) {
            if (mCameraRollFragment != null && mPublicPhotosFragment != null) {
                int color = ColorUtils.blendARGB(
                        backgroundColorCameraRoll, backgroundColorPublicPhotos, positionOffset);
                mCameraRollFragment.setBackgroundColor(color);
                mPublicPhotosFragment.setBackgroundColor(color);
            }
        } else if (position == 1) {
            if (mPublicPhotosFragment != null && mAlbumsFragment != null) {
                int color = ColorUtils.blendARGB(
                        backgroundColorPublicPhotos, backgroundColorAlbums, positionOffset);
                mPublicPhotosFragment.setBackgroundColor(color);
                mAlbumsFragment.setBackgroundColor(color);
            }
        } else if (position == 2) {
            if (mAlbumsFragment != null && mContactListFragment != null) {
                int color = ColorUtils.blendARGB(
                        backgroundColorAlbums, backgroundColorContactList, positionOffset);
                mAlbumsFragment.setBackgroundColor(color);
                mContactListFragment.setBackgroundColor(color);
            }
        }
    }

    private class MyFragmentPagerAdapter extends FragmentStatePagerAdapter {

        private MyFragmentPagerAdapter(FragmentManager fm) {
            super(fm);
        }

        @Override
        public Fragment getItem(int position) {
            switch (position) {
                case CAMERA_ROLL_FRAGMENT_ID:
                    mCameraRollFragment = CameraRollFragment.Companion.newInstance();
                    return mCameraRollFragment;
                case PUBLIC_PHOTOS_FRAGMENT_ID:
                    mPublicPhotosFragment = PublicPhotosFragment.newInstance();
                    return mPublicPhotosFragment;
                case ALBUMS_FRAGMENT_ID:
                    mAlbumsFragment = AlbumsFragment.newInstance();
                    return mAlbumsFragment;
                case CONTACT_LIST_FRAGMENT_ID:
                    mContactListFragment = ContactListFragment.newInstance();
                    return mContactListFragment;
                default:
                    mCameraRollFragment = CameraRollFragment.Companion.newInstance();
                    return mCameraRollFragment;
            }
        }

        @Override
        public int getCount() {
            return NUM_OF_PAGES;
        }

        @Override
        public CharSequence getPageTitle(int position) {
            switch (position) {
                case CAMERA_ROLL_FRAGMENT_ID:
                    return App.Companion.getAppContext().getString(R.string.camera_roll);
                case PUBLIC_PHOTOS_FRAGMENT_ID:
                    return App.Companion.getAppContext().getString(R.string.public_photos);
                case ALBUMS_FRAGMENT_ID:
                    return App.Companion.getAppContext().getString(R.string.albums);
                case CONTACT_LIST_FRAGMENT_ID:
                    return App.Companion.getAppContext().getString(R.string.contact_list);
                default:
                    return App.Companion.getAppContext().getString(R.string.camera_roll);
            }
        }

    }

}

File: app/src/main/java/com/mudrichenko/evgeniy/flickrtestproject/ui/profile/ProfilePresenter.kt
package com.mudrichenko.evgeniy.flickrtestproject.ui.profile

import androidx.core.content.ContextCompat
import androidx.core.graphics.ColorUtils
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentManager
import androidx.fragment.app.FragmentStatePagerAdapter
import androidx.viewpager.widget.PagerAdapter

import com.arellomobile.mvp.InjectViewState
import com.arellomobile.mvp.MvpPresenter
import com.mudrichenko.evgeniy.flickrtestproject.App
import com.mudrichenko.evgeniy.flickrtestproject.R
import com.mudrichenko.evgeniy.flickrtestproject.ui.albums.AlbumsFragment
import com.mudrichenko.evgeniy.flickrtestproject.ui.cameraRoll.CameraRollFragment
import com.mudrichenko.evgeniy.flickrtestproject.ui.contactList.ContactListFragment
import com.mudrichenko.evgeniy.flickrtestproject.ui.publicPhotos.PublicPhotosFragment

@InjectViewState
class ProfilePresenter : MvpPresenter<ProfileView>() {

    private val CAMERA_ROLL_FRAGMENT_ID = 0
    private val PUBLIC_PHOTOS_FRAGMENT_ID = 1
    private val ALBUMS_FRAGMENT_ID = 2
    private val CONTACT_LIST_FRAGMENT_ID = 3
    private val DEFAULT_FRAGMENT_ID = 0

    private val NUM_OF_PAGES = 4

    var viewPagerCurrentPosition: Int = 0

    private val backgroundColorCameraRoll: Int
    private val backgroundColorPublicPhotos: Int
    private val backgroundColorAlbums: Int
    private val backgroundColorContactList: Int

    private var mCameraRollFragment: CameraRollFragment? = null
    private var mPublicPhotosFragment: PublicPhotosFragment? = null
    private var mAlbumsFragment: AlbumsFragment? = null
    private var mContactListFragment: ContactListFragment? = null

    init {
        viewPagerCurrentPosition = DEFAULT_FRAGMENT_ID
        backgroundColorCameraRoll = ContextCompat.getColor(App.appContext!!, R.color.backgroundCameraRoll)
        backgroundColorPublicPhotos = ContextCompat.getColor(App.appContext!!, R.color.backgroundPublicPhotos)
        backgroundColorAlbums = ContextCompat.getColor(App.appContext!!, R.color.backgroundAlbums)
        backgroundColorContactList = ContextCompat.getColor(App.appContext!!, R.color.backgroundContactList)
    }

    fun getPagerAdapter(fragmentManager: FragmentManager): PagerAdapter {
        return MyFragmentPagerAdapter(fragmentManager)
    }

    fun onViewPagerScrolled(position: Int, positionOffset: Float) {
        if (position == 0) {
            if (mCameraRollFragment != null && mPublicPhotosFragment != null) {
                val color = ColorUtils.blendARGB(
                        backgroundColorCameraRoll, backgroundColorPublicPhotos, positionOffset)
                mCameraRollFragment!!.setBackgroundColor(color)
                mPublicPhotosFragment!!.setBackgroundColor(color)
            }
        } else if (position == 1) {
            if (mPublicPhotosFragment != null && mAlbumsFragment != null) {
                val color = ColorUtils.blendARGB(
                        backgroundColorPublicPhotos, backgroundColorAlbums, positionOffset)
                mPublicPhotosFragment!!.setBackgroundColor(color)
                mAlbumsFragment!!.setBackgroundColor(color)
            }
        } else if (position == 2) {
            if (mAlbumsFragment != null && mContactListFragment != null) {
                val color = ColorUtils.blendARGB(
                        backgroundColorAlbums, backgroundColorContactList, positionOffset)
                mAlbumsFragment!!.setBackgroundColor(color)
                mContactListFragment!!.setBackgroundColor(color)
            }
        }
    }

    private inner class MyFragmentPagerAdapter constructor(fm: FragmentManager) : FragmentStatePagerAdapter(fm) {

        override fun getItem(position: Int): Fragment {
            when (position) {
                CAMERA_ROLL_FRAGMENT_ID -> {
                    mCameraRollFragment = CameraRollFragment.newInstance()
                    return mCameraRollFragment as CameraRollFragment
                }
                PUBLIC_PHOTOS_FRAGMENT_ID -> {
                    mPublicPhotosFragment = PublicPhotosFragment.newInstance()
                    return mPublicPhotosFragment as PublicPhotosFragment
                }
                ALBUMS_FRAGMENT_ID -> {
                    mAlbumsFragment = AlbumsFragment.newInstance()
                    return mAlbumsFragment as AlbumsFragment
                }
                CONTACT_LIST_FRAGMENT_ID -> {
                    mContactListFragment = ContactListFragment.newInstance()
                    return mContactListFragment as ContactListFragment
                }
                else -> {
                    mCameraRollFragment = CameraRollFragment.newInstance()
                    return mCameraRollFragment as CameraRollFragment
                }
            }
        }

        override fun getCount(): Int {
            return NUM_OF_PAGES
        }

        override fun getPageTitle(position: Int): CharSequence? {
            when (position) {
                CAMERA_ROLL_FRAGMENT_ID -> return App.appContext!!.getString(R.string.camera_roll)
                PUBLIC_PHOTOS_FRAGMENT_ID -> return App.appContext!!.getString(R.string.public_photos)
                ALBUMS_FRAGMENT_ID -> return App.appContext!!.getString(R.string.albums)
                CONTACT_LIST_FRAGMENT_ID -> return App.appContext!!.getString(R.string.contact_list)
                else -> return App.appContext!!.getString(R.string.camera_roll)
            }
        }

    }

}
